File: openags/research/messaging/discord.py
"""Discord notification channel via Bot API / Webhook."""

from __future__ import annotations

import logging

import httpx

from openags.models import DiscordConfig

logger = logging.getLogger(__name__)

_LEVEL_COLOR = {
    "info": 0x3498DB,     # blue
    "warning": 0xF39C12,  # orange
    "error": 0xE74C3C,    # red
    "success": 0x2ECC71,  # green
}


class DiscordChannel:
    """Send notifications to a Discord channel via Bot API.

    Requires a bot token and channel_id.
    Alternatively, channel_id can be a webhook URL for simpler setup.
    """

    def __init__(self, config: DiscordConfig) -> None:
        self._token = config.token.get_secret_value()
        self._channel_id = config.channel_id
        self._base_url = "https://discord.com/api/v10"

    @property
    def channel_id(self) -> str:
        return "discord"

    async def send_text(self, text: str) -> bool:
        if self._channel_id.startswith("http"):
            return await self._send_webhook({"content": text})
        return await self._send_bot(text)

    async def send_markdown(self, markdown: str) -> bool:
        # Discord uses its own markdown format; wrap in code block for complex md
        return await self.send_text(markdown)

    async def send_notification(
        self,
        title: str,
        body: str,
        level: str = "info",
    ) -> bool:
        embed = {
            "title": title,
            "description": body,
            "color": _LEVEL_COLOR.get(level, 0x3498DB),
        }

        if self._channel_id.startswith("http"):
            return await self._send_webhook({"embeds": [embed]})
        return await self._send_bot_embed(embed)

    async def health_check(self) -> bool:
        try:
            async with httpx.AsyncClient(timeout=10) as client:
                r = await client.get(
                    f"{self._base_url}/users/@me",
                    headers={"Authorization": f"Bot {self._token}"},
                )
                return r.status_code == 200
        except Exception as e:
            logger.warning("Discord health check failed: %s", e)
            return False

    async def _send_bot(self, text: str) -> bool:
        """Send via Discord Bot API."""
        url = f"{self._base_url}/channels/{self._channel_id}/messages"
        headers = {"Authorization": f"Bot {self._token}"}

        try:
            async with httpx.AsyncClient(timeout=15) as client:
                r = await client.post(url, headers=headers, json={"content": text})
                if r.status_code in (200, 201):
                    return True
                logger.warning("Discord send failed: %s %s", r.status_code, r.text)
                return False
        except Exception as e:
            logger.error("Discord send error: %s", e)
            return False

    async def _send_bot_embed(self, embed: dict) -> bool:
        """Send an embed via Discord Bot API."""
        url = f"{self._base_url}/channels/{self._channel_id}/messages"
        headers = {"Authorization": f"Bot {self._token}"}

        try:
            async with httpx.AsyncClient(timeout=15) as client:
                r = await client.post(url, headers=headers, json={"embeds": [embed]})
                if r.status_code in (200, 201):
                    return True
                logger.warning("Discord embed failed: %s", r.text)
                return False
        except Exception as e:
            logger.error("Discord embed error: %s", e)
            return False

    async def _send_webhook(self, payload: dict) -> bool:
        """Send via Discord webhook URL."""
        try:
            async with httpx.AsyncClient(timeout=15) as client:
                r = await client.post(self._channel_id, json=payload)
                if r.status_code in (200, 204):
                    return True
                logger.warning("Discord webhook failed: %s", r.text)
                return False
        except Exception as e:
            logger.error("Discord webhook error: %s", e)
            return False

File: packages/app/src/messaging/discord.ts
/**
 * Discord Bot Integration
 *
 * Send messages via Discord webhooks or bot API.
 */

export interface DiscordConfig {
  /** Bot token for full API access */
  botToken?: string
  /** Webhook URL for simple messaging */
  webhookUrl?: string
  /** Default channel ID for notifications */
  channelId?: string
}

export interface DiscordEmbed {
  title?: string
  description?: string
  url?: string
  color?: number
  fields?: Array<{
    name: string
    value: string
    inline?: boolean
  }>
  footer?: {
    text: string
    icon_url?: string
  }
  timestamp?: string
}

export interface DiscordMessage {
  content?: string
  embeds?: DiscordEmbed[]
  username?: string
  avatar_url?: string
}

const DISCORD_API = 'https://discord.com/api/v10'

export class DiscordBot {
  private config: DiscordConfig

  constructor(config: DiscordConfig) {
    this.config = config
  }

  /**
   * Send a message via webhook.
   */
  async sendWebhook(message: DiscordMessage): Promise<boolean> {
    if (!this.config.webhookUrl) {
      throw new Error('webhookUrl not configured')
    }

    const response = await fetch(this.config.webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(message),
    })

    return response.ok
  }

  /**
   * Send a message to a channel via bot API.
   */
  async sendMessage(channelId: string, message: DiscordMessage): Promise<{ id: string } | null> {
    if (!this.config.botToken) {
      throw new Error('botToken not configured')
    }

    const url = `${DISCORD_API}/channels/${channelId}/messages`

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bot ${this.config.botToken}`,
      },
      body: JSON.stringify(message),
    })

    if (!response.ok) return null

    const result = await response.json() as { id: string }
    return result
  }

  /**
   * Send a notification to the default channel.
   */
  async notify(text: string, options?: { embed?: DiscordEmbed }): Promise<boolean> {
    // Prefer webhook if available
    if (this.config.webhookUrl) {
      return this.sendWebhook({
        content: text,
        embeds: options?.embed ? [options.embed] : undefined,
      })
    }

    // Fall back to bot API
    if (this.config.botToken && this.config.channelId) {
      const result = await this.sendMessage(this.config.channelId, {
        content: text,
        embeds: options?.embed ? [options.embed] : undefined,
      })
      return result !== null
    }

    throw new Error('No webhook URL or bot token configured')
  }

  /**
   * Send a rich embed notification.
   */
  async notifyEmbed(embed: DiscordEmbed): Promise<boolean> {
    return this.notify('', { embed })
  }

  /**
   * Create a research progress embed.
   */
  static createProgressEmbed(
    stage: string,
    status: 'running' | 'completed' | 'failed',
    details?: string
  ): DiscordEmbed {
    const colors = {
      running: 0x3498db,  // Blue
      completed: 0x2ecc71, // Green
      failed: 0xe74c3c,   // Red
    }

    return {
      title: `Research Stage: ${stage}`,
      description: details,
      color: colors[status],
      timestamp: new Date().toISOString(),
      footer: {
        text: `Status: ${status}`,
      },
    }
  }

  /**
   * Get bot info.
   */
  async getMe(): Promise<{ id: string; username: string } | null> {
    if (!this.config.botToken) {
      throw new Error('botToken not configured')
    }

    const url = `${DISCORD_API}/users/@me`

    const response = await fetch(url, {
      headers: {
        'Authorization': `Bot ${this.config.botToken}`,
      },
    })

    if (!response.ok) return null

    const result = await response.json() as { id: string; username: string }
    return result
  }

  /**
   * Get channel info.
   */
  async getChannel(channelId: string): Promise<{ id: string; name: string; type: number } | null> {
    if (!this.config.botToken) {
      throw new Error('botToken not configured')
    }

    const url = `${DISCORD_API}/channels/${channelId}`

    const response = await fetch(url, {
      headers: {
        'Authorization': `Bot ${this.config.botToken}`,
      },
    })

    if (!response.ok) return null

    const result = await response.json() as { id: string; name: string; type: number }
    return result
  }
}
